src/types/Writable.ts

```diff
@@ -15,6 +15,12 @@ type WritableKeys<T> = {
 	[P in keyof T]: IfEquals<{ [Q in P]: T[P] }, { -readonly [Q in P]: T[P] }, P>;
 }[keyof T];
 
+type OptionalKeys<T> = {
+	[P in keyof T]: {} extends Pick<T, P> ? P : never;
```

Comment:
This creates a [union type](https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#union-types).

A key `P` (why the name `P`? `K` or `Key` seems clearer here) is optional if `{} extends Pick<T, P>` - that is, if an empty object is a valid assignment for [`Pick<T, P>`](https://www.typescriptlang.org/docs/handbook/utility-types.html#picktype-keys), then the picked key must not be required.

So, `OptionalKeys<T>` is the union of all the optional keys.

Reply:
Given how magic and complex all this is, would it be worth adding some of this narrative in the form of comments?

Reply:
@slifty and I talked today, and worked out a refactoring that makes things clearer, and he plans to add more comments, so I'm marking this as resolved.

Reply:
I've added comments... do they make it clearer, who knows.